hs-bindgen/src/HsBindgen/C/Tc/Macro.hs

```diff
@@ -1231,93 +2048,114 @@ combineSubsts ( Subst s1 ) ( Subst s2 ) =
+    , instanceDefaults = mbDflt }
+    ) args
+    | Just Refl <- Vec.withDict args $ Nat.eqNat @nbArgs @instNbArgs
+    =
+  runTcGenMSubst do
+    let orig = ClassInstMetaOrigin $ Quant $ fmap ( fmap ( Class cls ) ) iqty
+    ( instBndrs, instArgTys ) <- instantiate ctOrig orig iqty
+    liftUnifyM $
+      unifyTypes ctOrig NotSwapped instArgTys args
+    matchSubst <- State.get
+    return $
+      mapMaybeA ( tryDefault ctOrig matchSubst . ( $ instBndrs ) ) mbDflt
+  | otherwise
+  = error $ unlines
+      [ "matchOneInst: incorrect class arity"
+      , "class: " ++ show cls
+      ]
+
+tryDefault :: CtOrigin -> Subst TyVar -> NE.NonEmpty ( Type Ty, Type Ty ) -> TcPureM ( Maybe ( Subst TyVar ) )
+tryDefault ctOrig matchSubst dfltEqs =
+  fmap ( fmap snd ) $ runTcUnifyMSubst matchSubst $
+    traverse ( uncurry $ unifyType ( DefaultingOrigin ctOrig ) NotSwapped ) dfltEqs
 
 {-------------------------------------------------------------------------------
-  Typechecking macros: generalisation
+  Typechecking macros: top-level entry point to constraint solving
 -------------------------------------------------------------------------------}
 
-tcMacro :: Map CName QuantTy -> CName -> [ CName ] -> MExpr -> Either TcMacroError QuantTy
-tcMacro tyEnv macroNm args body = do
+-- | Top-level type-checking monad.
+type TcTopM = ExceptT TcMacroError TcUniqueM
 
-  -- Step 1: infer the type.
-  let
-    ( ( ty, ctsOrigs ), mbErrs ) = runTcM tyEnv ( inferTop macroNm args body )
-  throwAnyErrs TcErrors mbErrs
+simplifyAndDefault :: IntSet -> Cts -> TcTopM ( Subst TyVar, Cts )
+simplifyAndDefault quantTvs cts =
+  do
+    ( (), ( subst, ( insols, inerts ) ) ) <- lift $ runTcSolveM cts $ solvingLoop solveOne
+    for_ ( NE.nonEmpty insols ) \ errs ->
+      Except.throwError ( TcInconsistentConstraints $ NE.singleton ( NE.toList errs ) )
+    return ( subst, inerts )
 
-  -- Step 2: compute the set of metavariables we will quantify over.
-  let
-    quantTvs = freeTyVarsOfType mempty ty
-
-  -- Step 3: simplify and default constraints.
-  simpleCts <- simplifyAndDefault ( tyVarSet quantTvs ) ctsOrigs
-
-  -- Step 4: generalise.
-  Vec.reifyList quantTvs $ \ ( quantTvsVec :: Vec n TyVar ) -> do
-    Right $ QuantTy @n $ \ tys ->
-      let quantSubst = mkSubst ( toList $ Vec.zipWith (,) quantTvsVec tys )
-      in QuantTyBody
-          { quantTyQuant = map ( applySubst quantSubst ) ( fmap fst simpleCts )
-          , quantTyBody  =       applySubst quantSubst   ty
-          }
+  where
+    solveOne = solveCt Defer ( DefaultTyVarsExcept quantTvs ) classInstancesWithDefaults
+
+{-------------------------------------------------------------------------------
+  Typechecking macros: generalisation
+-------------------------------------------------------------------------------}
+
+tcMacro :: Map CName ( Quant ( Type Ty ) ) -> CName -> [ CName ] -> MExpr -> Either TcMacroError ( Quant ( Type Ty ) )
```

Comment:
```suggestion
tcMacro :: Platform -> Map CName ( Quant ( Type Ty ) ) -> CName -> [ CName ] -> MExpr -> Either TcMacroError ( Quant ( Type Ty ) )
```

Reply:
This change will propagate but we don't have the infrastructure in place to pass this right now, so leave a TODO.